### ISSUE TITLE
AArch64 vfma_laneq_f16 uses dup instead of lane

### ISSUE THREAD

**Issue by fbarchard** [llvm/llvm-project#55254]
clang is not using lanes for vfma_laneq_f16 intrinsic

Instead of dup+fmla
```
   3df54: 73 00 c0 3d                  	ldr	q19, [x3]
   3df74: 64 06 02 0e                  	dup	v4.4h, v19.h[0]
   3df78: 65 06 0a 0e                  	dup	v5.4h, v19.h[2]
...
   3dff0: 9f 1c a4 4e                  	mov	v31.16b, v4.16b
   3dff4: 7f 0f 45 0e                  	fmla	v31.4h, v27.4h, v5.4h
```
Instead of fmla with lane
```
   3df54: 73 00 c0 3d                     ldr   q4, [x3]
..
   3df74: 64 06 02 0e                     dup   v31.4h, v4.h[0]
   3dff4: 7f 0f 45 0e                     fmla  v31.4h, v27.4h, v4.h[2]
```
This function: https://github.com/google/XNNPACK/blob/master/src/f16-dwconv2d-chw/gen/3x3p1-minmax-neonfp16arith-1x4.c
```
llvm-objdump -d --mcpu=cortex-a76 3x3p1-minmax-neonfp16arith-1x4.o

0000000000000000 <xnn_f16_dwconv2d_chw_ukernel_3x3p1__neonfp16arith_1x4>:
       0: 68 40 00 91   add     x8, x3, #16
       4: 15 e4 00 2f   movi    d21, #0000000000000000
       8: e2 20 40 fc   ldur    d2, [x7, #2]
       c: e3 00 40 fd   ldr     d3, [x7]
      10: 73 00 c0 3d   ldr     q19, [x3]
      14: 15 81 40 0d   ld1     { v21.s }[0], [x8]
      18: e0 0c 40 fd   ldr     d0, [x7, #24]
      1c: 29 1c 00 91   add     x9, x1, #7
      20: 01 e4 00 2f   movi    d1, #0000000000000000
      24: 28 f1 7d 92   and     x8, x9, #0xfffffffffffffff8
      28: 42 04 02 0e   dup     v2.4h, v2.h[0]
      2c: 63 04 02 0e   dup     v3.4h, v3.h[0]
      30: 64 06 02 0e   dup     v4.4h, v19.h[0]
      34: 65 06 0a 0e   dup     v5.4h, v19.h[2]
      38: 66 06 16 0e   dup     v6.4h, v19.h[5]
      3c: 67 06 06 0e   dup     v7.4h, v19.h[1]
      40: 70 06 12 0e   dup     v16.4h, v19.h[4]
      44: 71 06 1e 0e   dup     v17.4h, v19.h[7]
      48: 72 06 0e 0e   dup     v18.4h, v19.h[3]
      4c: 73 06 1a 0e   dup     v19.4h, v19.h[6]
      50: b4 06 02 0e   dup     v20.4h, v21.h[0]
      54: b5 06 06 0e   dup     v21.4h, v21.h[1]
      58: ea 03 04 aa   mov     x10, x4
      5c: 49 00 01 8b   add     x9, x2, x1
      60: 1f 08 00 f1   cmp     x0, #2
      64: 89 30 89 9a   csel    x9, x4, x9, lo
      68: 56 01 40 fd   ldr     d22, [x10]
      6c: 57 84 40 fc   ldr     d23, [x2], #8
      70: 38 85 40 fc   ldr     d24, [x9], #8
      74: 3f 24 00 f1   cmp     x1, #9
      78: 63 05 00 54   b.lo    0x124 <xnn_f16_dwconv2d_chw_ukernel_3x3p1__neonfp16arith_1x4+0x124>
      7c: eb 03 1f aa   mov     x11, xzr
      80: 4c 21 00 91   add     x12, x10, #8
      84: 1c e4 00 2f   movi    d28, #0000000000000000
      88: 1d e4 00 2f   movi    d29, #0000000000000000
      8c: 1e e4 00 2f   movi    d30, #0000000000000000
      90: ea 03 01 aa   mov     x10, x1
      94: db 1e b6 4e   mov     v27.16b, v22.16b
      98: 9f 1c a4 4e   mov     v31.16b, v4.16b
      9c: fa 1e b7 4e   mov     v26.16b, v23.16b
      a0: 7f 0f 45 0e   fmla    v31.4h, v27.4h, v5.4h
      a4: 19 1f b8 4e   mov     v25.16b, v24.16b
      a8: 96 69 6b fc   ldr     d22, [x12, x11]
      ac: 5f 0f 46 0e   fmla    v31.4h, v26.4h, v6.4h
      b0: 57 68 6b fc   ldr     d23, [x2, x11]
      b4: 9c 33 1b 2e   ext     v28.8b, v28.8b, v27.8b, #6
      b8: 9f 0e 59 0e   fmla    v31.4h, v20.4h, v25.4h
      bc: 38 69 6b fc   ldr     d24, [x9, x11]
      c0: bd 33 1a 2e   ext     v29.8b, v29.8b, v26.8b, #6
      c4: 9f 0f 47 0e   fmla    v31.4h, v28.4h, v7.4h
      c8: de 33 19 2e   ext     v30.8b, v30.8b, v25.8b, #6
      cc: bf 0f 50 0e   fmla    v31.4h, v29.4h, v16.4h
      d0: 7c 13 16 2e   ext     v28.8b, v27.8b, v22.8b, #2
      d4: df 0f 51 0e   fmla    v31.4h, v30.4h, v17.4h
      d8: 5d 13 17 2e   ext     v29.8b, v26.8b, v23.8b, #2
      dc: 9f 0f 52 0e   fmla    v31.4h, v28.4h, v18.4h
      e0: 3e 13 18 2e   ext     v30.8b, v25.8b, v24.8b, #2
      e4: bf 0f 53 0e   fmla    v31.4h, v29.4h, v19.4h
      e8: bf 0e 5e 0e   fmla    v31.4h, v21.4h, v30.4h
      ec: fc 37 43 0e   fmax    v28.4h, v31.4h, v3.4h
      f0: 4a 21 00 d1   sub     x10, x10, #8
      f4: 9c 37 c2 0e   fmin    v28.4h, v28.4h, v2.4h
      f8: 5f 21 00 f1   cmp     x10, #8
      fc: bc 68 2b fc   str     d28, [x5, x11]
     100: 6b 21 00 91   add     x11, x11, #8
     104: 7c 1f bb 4e   mov     v28.16b, v27.16b
     108: 5d 1f ba 4e   mov     v29.16b, v26.16b
     10c: 3e 1f b9 4e   mov     v30.16b, v25.16b
     110: 28 fc ff 54   b.hi    0x94 <xnn_f16_dwconv2d_chw_ukernel_3x3p1__neonfp16arith_1x4+0x94>
     114: 29 01 0b 8b   add     x9, x9, x11
     118: 42 00 0b 8b   add     x2, x2, x11
     11c: a5 00 0b 8b   add     x5, x5, x11
     120: 05 00 00 14   b       0x134 <xnn_f16_dwconv2d_chw_ukernel_3x3p1__neonfp16arith_1x4+0x134>
     124: ea 03 01 aa   mov     x10, x1
     128: 19 e4 00 2f   movi    d25, #0000000000000000
     12c: 1a e4 00 2f   movi    d26, #0000000000000000
     130: 1b e4 00 2f   movi    d27, #0000000000000000
     134: 16 1c 36 0e   and     v22.8b, v0.8b, v22.8b
     138: 9c 1c a4 4e   mov     v28.16b, v4.16b
     13c: 17 1c 37 0e   and     v23.8b, v0.8b, v23.8b
     140: dc 0e 45 0e   fmla    v28.4h, v22.4h, v5.4h
     144: 18 1c 38 0e   and     v24.8b, v0.8b, v24.8b
     148: fc 0e 46 0e   fmla    v28.4h, v23.4h, v6.4h
     14c: 7b 33 16 2e   ext     v27.8b, v27.8b, v22.8b, #6
     150: 9c 0e 58 0e   fmla    v28.4h, v20.4h, v24.4h
     154: 5a 33 17 2e   ext     v26.8b, v26.8b, v23.8b, #6
     158: 7c 0f 47 0e   fmla    v28.4h, v27.4h, v7.4h
     15c: 39 33 18 2e   ext     v25.8b, v25.8b, v24.8b, #6
     160: 5c 0f 50 0e   fmla    v28.4h, v26.4h, v16.4h
     164: d6 12 01 2e   ext     v22.8b, v22.8b, v1.8b, #2
     168: 3c 0f 51 0e   fmla    v28.4h, v25.4h, v17.4h
     16c: f7 12 01 2e   ext     v23.8b, v23.8b, v1.8b, #2
     170: dc 0e 52 0e   fmla    v28.4h, v22.4h, v18.4h
     174: 18 13 01 2e   ext     v24.8b, v24.8b, v1.8b, #2
     178: fc 0e 53 0e   fmla    v28.4h, v23.4h, v19.4h
     17c: bc 0e 58 0e   fmla    v28.4h, v21.4h, v24.4h
     180: 96 37 43 0e   fmax    v22.4h, v28.4h, v3.4h
     184: 5f 21 00 f1   cmp     x10, #8
     188: d6 36 c2 0e   fmin    v22.4h, v22.4h, v2.4h
     18c: e1 00 00 54   [b.ne](https://www.google.com/url?q=http://b.ne&sa=D&source=buganizer&usg=AOvVaw19sy2HnPy8XzNXzW9mLJje)    0x1a8 <xnn_f16_dwconv2d_chw_ukernel_3x3p1__neonfp16arith_1x4+0x1a8>
     190: b6 84 00 fc   str     d22, [x5], #8
     194: 4a 00 08 cb   sub     x10, x2, x8
     198: 00 04 00 f1   subs    x0, x0, #1
     19c: 22 01 08 cb   sub     x2, x9, x8
     1a0: e1 f5 ff 54   [b.ne](https://www.google.com/url?q=http://b.ne&sa=D&source=buganizer&usg=AOvVaw19sy2HnPy8XzNXzW9mLJje)    0x5c <xnn_f16_dwconv2d_chw_ukernel_3x3p1__neonfp16arith_1x4+0x5c>
     1a4: 07 00 00 14   b       0x1c0 <xnn_f16_dwconv2d_chw_ukernel_3x3p1__neonfp16arith_1x4+0x1c0>
     1a8: 6a 00 10 36   tbz     w10, #2, 0x1b4 <xnn_f16_dwconv2d_chw_ukernel_3x3p1__neonfp16arith_1x4+0x1b4>
     1ac: b6 80 9f 0d   st1     { v22.s }[0], [x5], #4
     1b0: d6 22 16 2e   ext     v22.8b, v22.8b, v22.8b, #4
     1b4: 0a ff 0f 36   tbz     w10, #1, 0x194 <xnn_f16_dwconv2d_chw_ukernel_3x3p1__neonfp16arith_1x4+0x194>
     1b8: b6 40 9f 0d   st1     { v22.h }[0], [x5], #2
     1bc: f6 ff ff 17   b       0x194 <xnn_f16_dwconv2d_chw_ukernel_3x3p1__neonfp16arith_1x4+0x194>
     1c0: c0 03 5f d6   ret
```
clang --version     
Android (7284624, based on r416183b) clang version 12.0.5 (https://android.googlesource.com/toolchain/llvm-project c935d99d7cf2016289302412d708641d52d2f7ee)
Target: x86_64-unknown-linux-gnu
Thread model: posix

**Comment by fbarchard** [comment]
The same issue affects vfma_laneq_f32 and vfma_lane_f16.

The issue appears to be an optimizer issue (-O2)
In different code the same intrinsic generates a lane.

Whats different about this code is a matrix of 3x3 or 5x5 is loaded into vectors outside the loop.
e.g.
  const float16x8_t vw01234567 = vld1q_f16(w0);
  const float16x8_t vw89ABCDEF = vld1q_f16(w0 + 8);
  const float16x8_t vwGHIJKLMN = vld1q_f16(w0 + 16);
  const float16x4_t vwOP = vreinterpret_f16_u32(vld1_lane_u32((const uint32_t*)(w0 + 24), vmov_n_u32(0), 0));

The dup's generate 9 or 25 vectors.  This causes register spill to the stack.
Plus additional push/pop of d8-d15 and up to 25 dups/stp to the stack

If I compile with -O0 lanes are used, but -O1, -O2, -O3, -Os and -Ofast all use dup
Is there a way to coax the code generation to use lanes?

**Comment by fhahn** [comment]
@fbarchard would it be possible to extract a standalone reproducer?

**Comment by fbarchard** [comment]
Re standalone.  I've attached a preprocessed file, using clang -E
[3x3p1-minmax-neonfp16arith-4x8.txt](https://github.com/llvm/llvm-project/files/9144762/3x3p1-minmax-neonfp16arith-4x8.txt)

**Comment by fhahn** [comment]
Thanks, there have been a number of related fixes since 12.x. With current `main` only lane variants for `fmla` are generated. Please verify this works as expected with a current version of Clang/LLVM (either a build from `main` or one of the upcoming release candidates for 15.0). If you still see issues, please re-open with more details.

```
bin/clang -arch arm64 -mcpu=cortex-a76 ~/Downloads/3x3p1-minmax-neonfp16arith-4x8.c -c -S -o - | grep fmla

	fmla.8h	v0, v1, v2[2]
	fmla.8h	v0, v1, v2[2]
	fmla.8h	v0, v1, v2[2]
	fmla.8h	v0, v1, v2[2]
	fmla.8h	v0, v1, v2[1]
	fmla.8h	v0, v1, v2[1]
	fmla.8h	v0, v1, v2[1]
	fmla.8h	v0, v1, v2[1]
	fmla.8h	v0, v1, v2[0]
	fmla.8h	v0, v1, v2[0]
	fmla.8h	v0, v1, v2[0]
	fmla.8h	v0, v1, v2[0]
	fmla.8h	v0, v1, v2[1]
	fmla.8h	v0, v1, v2[1]
	fmla.8h	v0, v1, v2[1]
	fmla.8h	v0, v1, v2[1]
	fmla.8h	v0, v1, v2[0]
	fmla.8h	v0, v1, v2[0]
	fmla.8h	v0, v1, v2[0]
	fmla.8h	v0, v1, v2[0]
	fmla.8h	v0, v1, v2[3]
	fmla.8h	v0, v1, v2[3]
	fmla.8h	v0, v1, v2[3]
	fmla.8h	v0, v1, v2[3]
	fmla.8h	v0, v1, v2[3]
	fmla.8h	v0, v1, v2[3]
	fmla.8h	v0, v1, v2[3]
	fmla.8h	v0, v1, v2[3]
	fmla.8h	v0, v1, v2[2]
	fmla.8h	v0, v1, v2[2]
	fmla.8h	v0, v1, v2[2]
	fmla.8h	v0, v1, v2[2]
	fmla.8h	v0, v1, v2[1]
	fmla.8h	v0, v1, v2[1]
	fmla.8h	v0, v1, v2[1]
	fmla.8h	v0, v1, v2[1]
	fmla.8h	v0, v1, v2[2]
	fmla.8h	v0, v1, v2[2]
	fmla.8h	v0, v1, v2[2]
	fmla.8h	v0, v1, v2[2]
	fmla.8h	v0, v1, v2[1]
	fmla.8h	v0, v1, v2[1]
	fmla.8h	v0, v1, v2[1]
	fmla.8h	v0, v1, v2[1]
	fmla.8h	v0, v1, v2[0]
	fmla.8h	v0, v1, v2[0]
	fmla.8h	v0, v1, v2[0]
	fmla.8h	v0, v1, v2[0]
	fmla.8h	v0, v1, v2[1]
	fmla.8h	v0, v1, v2[1]
	fmla.8h	v0, v1, v2[1]
	fmla.8h	v0, v1, v2[1]
	fmla.8h	v0, v1, v2[0]
	fmla.8h	v0, v1, v2[0]
	fmla.8h	v0, v1, v2[0]
	fmla.8h	v0, v1, v2[0]
	fmla.8h	v0, v1, v2[3]
	fmla.8h	v0, v1, v2[3]
	fmla.8h	v0, v1, v2[3]
	fmla.8h	v0, v1, v2[3]
	fmla.8h	v0, v1, v2[3]
	fmla.8h	v0, v1, v2[3]
	fmla.8h	v0, v1, v2[3]
	fmla.8h	v0, v1, v2[3]
	fmla.8h	v0, v1, v2[2]
	fmla.8h	v0, v1, v2[2]
	fmla.8h	v0, v1, v2[2]
	fmla.8h	v0, v1, v2[2]
	fmla.8h	v0, v1, v2[1]
	fmla.8h	v0, v1, v2[1]
	fmla.8h	v0, v1, v2[1]
	fmla.8h	v0, v1, v2[1]
```

**Comment by fbarchard** [comment]
You need to enable optimization to reproduce the issue

The issue still occurs on clang 15
./clang --version
Android (8850317, based on r458507) clang version 15.0.1 (https://android.googlesource.com/toolchain/llvm-project 640d06f2a4ef16626a747d4c45f9bd9a9fdcef4c)
Target: x86_64-unknown-linux-gnu
Thread model: posix

./clang --target=aarch64-linux-android21 -O1 -arch arm64 -mcpu=cortex-a76 ./3x3p1-minmax-neonfp16arith-4x8.c -c -S -o - | grep fmla
	fmla	v10.8h, v6.8h, v24.8h
	fmla	v11.8h, v6.8h, v1.8h
	fmla	v12.8h, v6.8h, v31.8h
	fmla	v13.8h, v6.8h, v21.8h
	fmla	v10.8h, v7.8h, v1.8h
	fmla	v11.8h, v7.8h, v31.8h
	fmla	v12.8h, v7.8h, v21.8h
	fmla	v13.8h, v7.8h, v3.8h
	fmla	v10.8h, v2.8h, v31.8h
	fmla	v10.8h, v16.8h, v3.8h
	fmla	v11.8h, v2.8h, v21.8h
	fmla	v11.8h, v16.8h, v3.8h
	fmla	v10.8h, v18.8h, v3.8h
	fmla	v12.8h, v2.8h, v29.8h
	fmla	v13.8h, v2.8h, v28.8h
	fmla	v12.8h, v16.8h, v3.8h
	fmla	v13.8h, v16.8h, v22.8h
	fmla	v11.8h, v18.8h, v3.8h
	fmla	v10.8h, v0.8h, v3.8h
	fmla	v12.8h, v18.8h, v22.8h
	fmla	v13.8h, v18.8h, v25.8h
	fmla	v11.8h, v0.8h, v22.8h
	fmla	v12.8h, v0.8h, v25.8h
	fmla	v13.8h, v0.8h, v26.8h
	fmla	v10.8h, v26.8h, v22.8h
	fmla	v11.8h, v26.8h, v25.8h
	fmla	v10.8h, v19.8h, v25.8h
	fmla	v12.8h, v26.8h, v22.8h
	fmla	v11.8h, v19.8h, v22.8h
	fmla	v10.8h, v5.8h, v22.8h
	fmla	v13.8h, v26.8h, v25.8h
	fmla	v12.8h, v19.8h, v25.8h
	fmla	v11.8h, v5.8h, v25.8h
	fmla	v13.8h, v19.8h, v22.8h
	fmla	v12.8h, v5.8h, v22.8h
	fmla	v13.8h, v5.8h, v25.8h
	fmla	v0.8h, v10.8h, v22.8h
	fmla	v1.8h, v10.8h, v25.8h
	fmla	v28.8h, v10.8h, v26.8h
	fmla	v3.8h, v10.8h, v21.8h
	fmla	v0.8h, v6.8h, v25.8h
	fmla	v1.8h, v6.8h, v26.8h
	fmla	v28.8h, v6.8h, v21.8h
	fmla	v3.8h, v6.8h, v30.8h
	fmla	v0.8h, v16.8h, v26.8h
	fmla	v1.8h, v16.8h, v21.8h
	fmla	v28.8h, v16.8h, v30.8h
	fmla	v3.8h, v16.8h, v31.8h
	fmla	v0.8h, v7.8h, v24.8h
	fmla	v1.8h, v7.8h, v27.8h
	fmla	v28.8h, v7.8h, v23.8h
	fmla	v3.8h, v7.8h, v9.8h
	fmla	v0.8h, v2.8h, v27.8h
	fmla	v1.8h, v2.8h, v23.8h
	fmla	v28.8h, v2.8h, v9.8h
	fmla	v3.8h, v2.8h, v29.8h
	fmla	v0.8h, v17.8h, v23.8h
	fmla	v1.8h, v17.8h, v9.8h
	fmla	v28.8h, v17.8h, v29.8h
	fmla	v3.8h, v17.8h, v8.8h
	fmla	v0.8h, v18.8h, v22.8h
	fmla	v1.8h, v18.8h, v23.8h
	fmla	v28.8h, v18.8h, v24.8h
	fmla	v3.8h, v18.8h, v21.8h
	fmla	v0.8h, v19.8h, v23.8h
	fmla	v1.8h, v19.8h, v24.8h
	fmla	v28.8h, v19.8h, v21.8h
	fmla	v3.8h, v19.8h, v25.8h
	fmla	v0.8h, v5.8h, v24.8h
	fmla	v1.8h, v5.8h, v21.8h
	fmla	v28.8h, v5.8h, v25.8h
	fmla	v3.8h, v5.8h, v26.8h

**Comment by davemgreen** [comment]
This version seems to be doing fine at all the optimization levels I tried:
https://godbolt.org/z/6o3rPcKz8

**Comment by fbarchard** [comment]
Agreed, that looks correct on godbolt
I tried godbolt with clang 14, and the dup is used, indicating trunk has fixed the issue.

unclear what the difference is between godbolt and clang 15 I tried
I tried the newest version I could, which is Android (8850317, based on r458507) clang version 15.0.1. godbolt is on 15.0.0
I wondered if its a clang c vs clang++ c++ difference, but it doesnt seem to be.
I tried my build options on godbolt, but lane is still used correctly.
next issue will be getting this to build with aarch32.
